### PR TITLE
Addressing a few issues regarding the OP reentry scenarios

### PR DESCRIPTION
**Pull request by andre4i**
## Description

Related to ADO:4537.

[Grouped batching](https://github.com/microsoft/FluidFramework/tree/main/packages/runtime/container-runtime/src/opLifecycle#grouped-batching) is problematic in the context of op reentry, as all ops within the same batch have the same sequence number. A reentrant op (created during the execution of a 'changed' event callback of a DDS) will have a different baseline than the rest of the batch and with the erasure of sequence numbers, it is impossible for the DDSes to properly apply the op. This PR addresses this issue by detecting batches with reentrant ops and rebasing them. Rebasing a batch means resubmitting it to the runtime to have all its ops in agreement re reference sequence numbers and sequence numbers. The issue is showcased in the `Eventual consistency with op reentry` end to end test, which is currently disabled until the DDS supports rebasing, which will happen very soon. The test proves that enabling group batching risks creating inconsistent data models in the op reentry scenario.

In a nutshell, this change will force the offending client to reorder its batch before it is sent, ensuring that all clients have the same view of the data model.

### Changes included in this PR:

- allow for rebasing ops, opt-in (disabled by default) via runtime options with a feature gate override for shutting it down.
- increase the captured stack length whenever we detect op reentry. This will be useful to identify problematic customer scenarios to leave our options open to possibly disable op reentry altogether in the future.
- flushing is blocked inside DDS event handlers. This is to simplify the possible states in which the runtime can end up and also ensure that a client cannot spam the backend with ops in case of unbounded recursion.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -435,9 +436,24 @@ export interface IContainerRuntimeOptions {
...
+	 * when an op is created as the result of processing another op. Usually this is the case
+	 * when changes are made to a DDS inside a DDS 'onChanged' event handler. This means that the
+	 * reentrant op will have a different reference sequence number  than the rest of the ops in
+	 * the batch, resulting in  a different view of the state of the data model. Therefore all ops
+	 * must be resubmitted and rebased to the current reference sequence number to be in agreement
+	 * about the state of the data model.
+	 *
+	 * @experimental Not ready for use.
+	 */
+	readonly enableBatchRebasing?: boolean;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:456]: Let's open a task to remove this flag in some distant future - this feels short lived (the op grouping option has much longer time horizon)

File: packages/runtime/container-runtime/src/opLifecycle/batchManager.ts [hunk 2]
```diff
@@ -54,9 +55,14 @@ export class BatchManager {
 
 	constructor(public readonly options: IBatchManagerOptions) {}
 
-	public push(message: BatchMessage, currentClientSequenceNumber?: number): boolean {
+	public push(
+		message: BatchMessage,
+		currentClientSequenceNumber?: number,
+		reentrant: boolean = false,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/batchManager.ts:61]: Does it need to be optional?

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 3]
```diff
@@ -191,22 +219,85 @@ export class Outbox {
...
 		) {
-			this.flushInternal(this.attachFlowBatch.popBatch());
+			this.flushInternal(this.attachFlowBatch);
 		}
 	}
 
 	public flush() {
-		this.flushInternal(this.attachFlowBatch.popBatch());
-		this.flushInternal(this.mainBatch.popBatch());
+		if (this.isContextReentrant()) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:227]: Should this be an assert? It might make sense (from code size POV) if we expect to never get here, and also if we get here, it (exception) is not lost and bubbles up somehow into logs.
If that removes the need to have one extra callback (closeContainer) - even better.

**Review by vladsud — APPROVED** [review]

**Review by kian-thompson — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:456]: ADO:4829

File: packages/runtime/container-runtime/src/opLifecycle/batchManager.ts (hunk 2 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/batchManager.ts:61]: Good point. This is the result of my laziness to not have to change all unit tests but it does need to be required, as a forcing function that if/when these things get refactored, we maintain correctness.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 3 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:227]: An assert here doesn't bubble up in the logs. It depends on the client handling it, this bubbles up to the event handler so it will be somehow hidden.

It is technically possible to get there with immediate flush and calling ordersequentially. The thing is that I'd like for this to close the container as recursion in this context can create a small storm of ops if flushing happens and this should preferably be as loud as possible.
